### DISCUSSION TITLE
Simultaneous input and output Teensy 4.1 -> Audio Shield

### DISCUSSION THREAD

**Discussion by Harrolee** [micropython/micropython#9939] [Core Development] [answered]
Hello!

Does micropython support (near) simultaneous input and output on the Teensy 4.1 -> pjrc Audio Shield? If not, how could I help support that?

Paul Stoffregen seems to have figured it out. His Teensy Audio library contains an example of simultaneous read-write: [PassThroughStereo.ino](https://github.com/PaulStoffregen/Audio/blob/master/examples/HardwareTesting/PassThroughStereo/PassThroughStereo.ino). Also, apparently it was broken for a time and then fixed in [this commit](https://github.com/PaulStoffregen/Audio/commit/bd5332f9d9a6c3095c0563dacafd086b37474443?diff=split). 

Paul's implementation uses either...
1. ...a single I2S bus for i/o
2. ...two busses, leveraging the same SCK and WC signals for both Tx and Rx to communicate with the [SGTL5000] (https://www.nxp.com/products/audio-and-radio/audio-converters/ultra-low-power-audio-codec:SGTL5000), which has only one SCK and WC input (thanks for this [technical explanation](https://github.com/orgs/micropython/discussions/9243#discussioncomment-3597153), @miketeachman !)

If Paul's IS2 objects (ie: [AudioOutputI2S](https://github.com/PaulStoffregen/Audio/blob/master/output_i2s.cpp) and [AudioInputI2S.cpp](https://github.com/PaulStoffregen/Audio/blob/master/input_i2s.cpp)) use a single I2S bus, could MicroPython's I2S class be changed support something similar? Having looked at those files and their parent, [AudioStream](https://github.com/PaulStoffregen/cores/blob/master/teensy4/AudioStream.cpp), I can't tell whether each IS2 object is its own I2S bus or if somehow, cleverly, requests from both instances of the AudioStream class are handled by a single I2S bus. I don't know C++ well enough to know where to look.

If Paul uses two I2S busses with the same SCK and WS pins, how could we change MicroPython's I2S class or @rdagger 's [SGTL5000 class](https://github.com/rdagger/micropython-sgtl5000) to allow using the same SCK and WS pins for two IS2 instances? I find that I get an EPERM error when I try to use two I2S instances that use the same SCK and WS pins. 

Thank you! 

related question: [simultaneous i/o for the ESP32](https://github.com/orgs/micropython/discussions/9243#discussion-4372459)
related thread: [Teensy 4.0 & 4.1 forum, SGTL5000 conversation](https://forum.micropython.org/viewtopic.php?f=12&t=6783&start=210#p64971)

**Comment by miketeachman** [comment, marked as answer]
> Does micropython support (near) simultaneous input and output on the Teensy 4.1 -> pjrc Audio Shield

I've never tried to do this, but it might be possible. 
Here are two approaches I can suggest:

1. Configure two I2S instances, one for receive, one for transmit.  To avoid the SCK/WS EPERM error, map the 2nd I2S instance to unused SCK/WS pins.  The pin choices for Teensy I2S pins are listed [here](https://github.com/miketeachman/micropython-i2s-examples#mimxrt-gpio-mappings-for-sck-ws-sd-mck).  Use an oscilloscope to determine if the SCK and WS signals from the two I2S instances are lined up.  If the signals line up you will be able to simultaneously receive and transmit audio samples from the codec on the Teensy audio shield.  However, I doubt that this approach will work as I suspect that each I2S instance will behave independently, creating clock signals that are not synchronized.  But it is definitely worth trying before moving onto a more complicated work around.  You might get lucky !

2. Understand how the Teensy audio library implements the simultaneous receive and transmit of audio samples from the codec.  Get a deep understanding of how it works.  Then apply this knowledge to attempt a workaround using machine.memX() functions to modify the low-level configuration registers in the the SAI peripherals (the SAI peripherals implement the I2S protocol).  The idea is to start with approach 1 above, then add micropython machine.memX() code to directly write the two SAI peripheral configurations until the WS and SCK signals are synchronous.  You will need to download a copy of NXP's `i.MX RT1064 Processor Reference Manual` to get the memory addresses of the SAI peripherals.  This may or may not be possible.  If you can figure this out you will definitely be crowned a micropython I2S ninja !

**Reply by Harrolee** [reply]
Thank you Mike! I'll give it a go.

**Reply by Harrolee** [reply]
I think the permissions error is thrown whenever one I2S instance tries to use a pin registered to a different I2S instance.
After changing the SCK/WS pins, I got the same error, likely because both I2S instances share a MCK pin 😝. I was able to avoid the EPERM error by using the SCK/WS/MCK pins for I2S id 2 (Thank you for linking that chart!). This led me to realize that the pjrc Audio Library implementation must use the same MCK for both busses (if it uses two busses) because I have nothing connected to the SCK/WS/MCK pads for I2S id 2.
